aten/src/ATen/native/cuda/EmbeddingBag.cu

```diff
@@ -62,43 +63,44 @@ __global__ void EmbeddingBag_updateOutputKernel(
       int64_t bag_size_ = 0;
       int64_t maxWord = -1;
       for (int64_t emb = begin; emb < end; emb++) {
+        bool pad = input[emb] == padding_idx;
         const int64_t weightRow = input[emb] * weight_stride0;
-        scalar_t weightValue = weightFeat[weightRow];
-
-        if (mode == MODE_MAX) {
-          if (emb == begin || weightValue > weightFeatMax) {
-            weightFeatMax = weightValue;
-            maxWord = input[emb];
-          }
-        } else {
-          if (per_sample_weights) {
-            accscalar_t scaleWeightBy = static_cast<accscalar_t>(
-                per_sample_weights[emb * per_sample_weights_stride]);
-            weightFeatSum += scaleWeightBy * static_cast<accscalar_t>(weightValue);
+        scalar_t weightValue = pad ? static_cast<scalar_t>(0) : weightFeat[weightRow];
+        if (!pad) {
```

Comment:
I wonder if performance will be affected by warp divergence?

Reply:
Yeah you're probably right. Maybe it would be best to remove this `if`, and use ternary operators in the body to make it an effective no-op if `pad == false`, kind of like this:

```
if (mode == MODE_MAX) {
  if (bag_size_ == 0 || weightValue > weightFeatMax) {
    weightFeatMax = pad ? weightFeatMax : weightValue;
    maxWord = pad ? maxWord : input[emb];
  }
} else {
  if (per_sample_weights) {
    accscalar_t scaleWeightBy = static_cast<accscalar_t>(
        per_sample_weights[emb * per_sample_weights_stride]);
    weightFeatSum += scaleWeightBy * static_cast<accscalar_t>(pad ? 0 : weightValue);
  } else {
    weightFeatSum += static_cast<accscalar_t>(pad ? 0 :weightValue);
  }
}
bag_size_ += pad ? 0 : 1;
```

Reply:
I wrote a benchmark script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/embedding_bag-padding_idx/embedding_bag-perf.py

You were right! Performance was down by roughly 30% in the worst case for CUDA, even with `padding_idx=None`.

With the above change, it improves to a drop of about 18%. I'll look for more improvements to see if I can bring it closer to 0%.

Reply:
I've improved performance for CUDA and CPU, back up to what it used to be before this PR. Look below for a summary of performance